### ISSUE TITLE
RFC: Splitting this repo up

### ISSUE THREAD

**Issue by droyad** [DbUp/DbUp#431]
I would like to get feedback on whether splitting this repo up so that each database provider has it's own repo and issue list is a good idea. I'm thinking something like [serilog](https://github.com/serilog/serilog) and it's [sinks](https://github.com/serilog/serilog-sinks-console) and [extensions](https://github.com/serilog/serilog-extensions-hosting).

Pros:

- Each provider can be versioned at it's own pace. For example, #365 requires a `major` bump from my reading of SemVer. It's be nice to do that just for the Firebird provider.
- We can have different maintainers for each database type, few people have experience across all of them. Those maintainers then also then only see issues relevant to their provider. 
- GH now allows transfers of issues, which makes the "wrong repo" thing less of an issue
- It forces work on the core to be isolated from the provider
- When making changes to core it doesn't need to be tested on all providers. This is prohibitive at the moment as it requires knowledge and an install of all of them. We could instead update the most popular ones and bring the rest along periodically.

Cons:
- The provider and core `major` versions may not match, which could cause confusion
- It's more juggling for someone working on the core and providers at the same time

I'm contemplating v5.0.0 as a major change to the project mainly to support extra script/`schemaversions` metadata and make code migrations first class. Each provider would be split out when it supported the changes in 5.0.0.

**Comment by AdrianJSClark** [comment]
I feel like this is the future Jake was working towards by splitting up the core and SQL Server code.

Regarding the potential version mismatch between the providers and core, I'd suggest that modern world users of the libraries shouldn't ever see the core package anyway. They should just install `DbUp.SqlServer` or whatever their chosen database provider is and let NuGet resolve the `DbUp.Core` dependency for them.

You mention it is more juggling for someone working on core and providers. Perhaps that's a good thing because it will expose where there aren't great seams or there are types assigned to the wrong place. So that friction might assist in the core truly containing core, shared technology.

TL;DR - sounds great!

**Comment by droyad** [comment]
@mjauernig  I'm back on the side of splitting it up after having looked through the issue list. It's mainly for the above reasons. I and the other maintainers have hit on in the past.

I think _not_ having separate repos causes friction for us and end users:

-  If we want to use the next major of say SqlClient we either have to:
    - Wait until the next major cut of DbUp
    - Do it immediately and all user's not using the SQL Provider get a no-op major bump
- If we want to make a minor change to a provider say Oracle, we have to bump the minor for _everyone_. Again, most people will have to do a no-op upgrade.

We should be aiming to ship as fast as possible, these factors I think will lead to us delaying the rollout of changes for months at a time. 

A Monorepo can work if we are able to version each provider independently.

**Comment by droyad** [comment]
If we are going to drop mainline support for a provider, it'd also be good to have a repo that we can hotfix security patches into (sure we could do that with branches too).

**Comment by droyad** [comment]
Since there is no further comments on this plan, and agreement from most current and former maintainers with experience with the problems described, I am going to start the work to make this happen.

**Comment by mjauernig** [comment]
Sorry @droyad for not answering earlier. I think splitting the repo is not a good idea. 

Main reason is that we do not have active maintainers for the different databases. I think splitting the repo will result in one active mssql branch, maybe postgresql will be active too. The rest I think will die out over time. There will certainly be some feature and pull requests from time to time for the other databases, but no active development.

As an example. In a year DbUp.Core + DbUp.MSSQL is on v7. DbUp.Oracle is on v5 and there is a pull request for DbUp.Oracle based on v5. We can surely integrate them. But who will check if the merge fits with the v7 of DbUp.Core. Maybe the merge will introduce some changes that will conflict with the v7 development.

In the end we will have different DbUps for different databases with different features.

I would change my mind if we had minimum two active maintainers for each database, who care about _being on track_.

The next point is "the need". At the moment (also due to the lack of too few maintainers) there are no core changes or development progress on the horizon. So we are solving a "problem" for a currently non-existent problem. Right now I see more pressing issues like writing good documentation and providing support on the issue board. 

And at least we have a more complex world to maintain. We have to develop 7-8 build pipelines and take care of the integrations. And there are a lot of sidesteps. Simple example: When someone asks a question about a specific database, we have to think about what features were in that database. 

I understand the problem with the version bump, but I would live with it. Because DbUp depends on database libraries and other frameworks that also do frequent version bumps, which will force us to do the same. Maybe we can solve the problem via the build pipeline. If you like, I can check if this is possible.

Personally, I switched my repositories to monorepositories half a year ago, and it has saved me a lot of non-productive time.
